### PR TITLE
Fix bug with projections and zero-copy replication

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix bug which can lead to broken projections if zero-copy replication is enabled and used.

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -6412,7 +6412,25 @@ std::pair<MergeTreeData::MutableDataPartPtr, scope_guard> MergeTreeData::cloneAn
...
+            {
+                auto file_name_with_projection_prefix = fs::path(projection_storage.getPartDirectory()) / it->name();
+                if (!files_to_copy_instead_of_hardlinks.contains(file_name_with_projection_prefix)
+                    && it->name() != IMergeTreeDataPart::DELETE_ON_DESTROY_MARKER_FILE_NAME
+                    && it->name() != IMergeTreeDataPart::TXN_VERSION_METADATA_FILE_NAME)
+                {
+                    hardlinked_files->hardlinks_from_source_part.insert(file_name_with_projection_prefix);
+                }
+            }
         }
```
Comment (CurtizJ) [inline src/Storages/MergeTree/MergeTreeData.cpp:6434]: Do we need to do the same [here](https://github.com/ClickHouse/ClickHouse/blob/150ba40b149d62f61ea8ebbf10c9fd3a5a002f4e/src/Storages/MergeTree/MutateTask.cpp#L1315-L1330)?

Comment (alesapin) [reply src/Storages/MergeTree/MergeTreeData.cpp:6434]: Absolutely

**Comment by alesapin** [comment]
BTW, no idea why test don't reproduce the bug. On my machine it reproduces.

**Comment by alesapin** [comment]
No related failures.

**Pull request merged**